### PR TITLE
[WIP] Refactor event, output files and export.to_obspy

### PR DESCRIPTION
**Pull request by twinder36**
This PR aims to tidy and internalise the handling of event attributes, review the .event file output, and set floating point precision here and in the .picks and .amps outputs. Scripts in the `QMigrate.io.export` submodule will then be modified accordingly.

Some of these changes are required to be made in tandem with changing / adding attributes to the LUT module, which is also under review. As such, some issues raised here will be have to be resolved in coordination with those changes.

**Aims**
- [x] Internalise event dataframe to the `Event.write()` method; this includes correcting access to some event attributes elsewhere.
 * This has been covered in https://github.com/QuakeMigrate/QuakeMigrate/commit/b969d0b9b016b2533cab1f1a698ce1291aae4226
- [x] Specify precision for floats in the .event, .amps and .picks files.
 * Covered in https://github.com/QuakeMigrate/QuakeMigrate/commit/739a47881b021957c6cfe11a620cbfa50a5312cc , however the current implementation is inefficient; these precisions should be calculated once for the LUT and stored as an attribute there, rather than once for every event.
- [x] Review the .event file format: which outputs are needed / useful.
 * Covered in https://github.com/QuakeMigrate/QuakeMigrate/pull/76/commits/739a47881b021957c6cfe11a620cbfa50a5312cc
- [x] Review the `QMigrate.io.export` submodule: make modifications for compatibility with new .event file format, and handling of Event attributes; add magnitudes functionality.
* In progress. `to_obspy()` method is covered in https://github.com/QuakeMigrate/QuakeMigrate/pull/76/commits/88768c038a8aaa5fe482bd6fc4cba4452a4f1ea2
* Other export functions will remain a WIP: this will be made clear in the documentation ahead of pushing to master.
- [x] Review soft deprecation for the old .event file format, mainly with respect to the `QMigrate.io.export` submodule
* This will not be included. Users can edit old .event files by simply renaming the columns as specified in https://github.com/QuakeMigrate/QuakeMigrate/pull/76/commits/739a47881b021957c6cfe11a620cbfa50a5312cc

**Improvements still required**
- [x] More efficient implementation of the function to determine the minimum coordinate precision for location outputs in the .event file. Move to LUT. But also ideally maintain backwards compatibility (this may not be possible).
* This will be included in #75 and the current implementation replaced after merging.

**List of major (API-affecting) changes**
* Modified the format of the .event files

### CODE REVIEW

File: QMigrate/io/event.py [hunk 1]
```diff
@@ -387,19 +384,57 @@ def write(self, run):
...
+            if axis == "Z":
+                decimals = max((dec_pos * -1 + 2), 0)
+                cols.extend(event_df.filter(regex="Err[X,Y,Z]"))
+            else:
+                decimals = max((dec_pos * -1 + 2), 6)
+            for col in cols:
+                event_df[col] = event_df.loc[:, col].round(decimals=decimals)
+                if decimals <= 0:
+                    event_df[col] = event_df.loc[:, col].astype(int)
+
```
Comment (hemmelig) [inline QMigrate/io/event.py:428]: Flagging this as a reminder to add a property to the LUT class that returns `dec_pos`, something like

```python
dec_pos = lut.precision
...
   class LUT:
        ...
        @property
        def precision(self):
            """
            Get appropriate number of decimal places as a function of the
            cell size and coordinate projection.
            """

            ll_corner = self.coord2grid(self.ll_corner, inverse=True)
            one_in = lut.index2coord([1, 1, 1])
            return [int(np.format_float_scientific(axis).split("e")[1])
                    for axis in (ll_corner - one_in)[0]]
```

File: QMigrate/signal/local_mag/magnitude.py [hunk 2]
```diff
@@ -386,14 +385,11 @@ def plot_amplitudes(self, magnitudes, event, run, noise_measure="RMS"):
         mag_err = event.localmag["ML_Err"]
         mag_r2 = event.localmag["ML_r2"]
 
-        # Calculate distance error (for errorbars)
-        gauss_loc = event.locations["gaussian"]
-        x_err = gauss_loc["LocalGaussian_ErrX"] / 1000
-        y_err = gauss_loc["LocalGaussian_ErrY"] / 1000
+        # Calculate distance error (for errorbars - using gaussian uncertainty)
+        x_err, y_err, z_err = event.get_loc_uncertainty("gaussian") / 1000
```
Comment (hemmelig) [inline QMigrate/signal/local_mag/magnitude.py:389]: Flag to remember this

File: QMigrate/io/event.py (hunk 1 above)
Comment (twinder36) [reply QMigrate/io/event.py:428]: This looks good. The current implementation will be retained for this PR, but can be replaced with this more efficient method after merging changes from #75

Note: for precision, the value should be multiplied by -1:

```
            return [int(np.format_float_scientific(axis).split("e")[1]) for axis
                    in np.subtract(*self.index2coord([[0, 0, 0], [1, 1, 1]]))] * -1
```

Comment (hemmelig) [reply QMigrate/io/event.py:428]: Good catch, added to my branch

File: QMigrate/io/export/to_obspy.py [hunk 3]
```diff
@@ -11,112 +11,254 @@
...
+    # Add COA info to extra
+    event.extra.coa = event_info["COA"]
+    event.extra.coa_norm = event_info["COA_NORM"]
+    event.extra.trig_coa = event_info["TRIG_COA"]
+    event.extra.dec_coa = event_info["DEC_COA"]
+    event.extra.dec_coa_norm = event_info["DEC_COA_NORM"]
+
+    # Determine location of cut waveform data - add to event object as a
+    # custom extra attribute.
+    mseed = locate_dir / "cut_waveforms" / "{}".format(event_file.stem)
```
Comment (hemmelig) [inline QMigrate/io/export/to_obspy.py:131]:
```suggestion
    mseed = locate_dir / "cut_waveforms" / event_file.stem
```

File: QMigrate/io/export/to_obspy.py [hunk 4]
```diff
@@ -11,112 +11,254 @@
...
+        origin.depth_errors.uncertainty = event_info["GAU_ErrZ"] * factor
+        origin.origin_uncertainty = ouc
+
+    # Add OriginQuality info to each origin?
+    for origin in event.origins:
+        origin.origin_type = "hypocenter"
+        origin.evaluation_mode = "automatic"
+
+    # --- Handle picks file ---
+    pick_file = locate_dir / "picks" / "{}".format(event_file.stem)
```
Comment (hemmelig) [inline QMigrate/io/export/to_obspy.py:179]:
```suggestion
    pick_file = locate_dir / "picks" / event.resource_id
```

File: QMigrate/io/export/to_obspy.py (hunk 3 above)
Comment (twinder36) [reply QMigrate/io/export/to_obspy.py:131]:
```suggestion
    mseed = locate_dir / "cut_waveforms" / event.resource_id.id
```

File: QMigrate/io/export/to_obspy.py (hunk 4 above)
Comment (twinder36) [reply QMigrate/io/export/to_obspy.py:179]:
```suggestion
    pick_file = locate_dir / "picks" / event.resource_id.id
```

File: QMigrate/io/export/to_obspy.py [hunk 5]
```diff
@@ -11,112 +11,254 @@
...
+                pick.extra.snr = float(pickline["SNR"])
+            elif method == "modelled":
+                pick.time = UTCDateTime(pickline["ModelledTime"])
+            else:
+                continue
             event.picks.append(pick)
-            arrival.pick_id = pick.resource_id
+
+    # --- Handle amplitudes file ---
+    amps_file = locate_dir / "amplitudes" / "{}".format(event_file.stem)
```
Comment (twinder36) [inline QMigrate/io/export/to_obspy.py:207]:
```suggestion
    amps_file = locate_dir / "amplitudes" / event.resource_id.id
```

File: QMigrate/io/export/to_obspy.py [hunk 6]
```diff
@@ -11,112 +11,254 @@
...
+    -------
+    event : `obspy.Event` object
+        Event object populated with all available information output by
+        `QMigrate.signal.scan.locate()`, including event locations and
+        uncertainties, picks, and amplitudes and magnitudes if available.
 
     """
 
     # Parse information from event file
     event_info = pd.read_csv(event_file).iloc[0]
```
Comment (twinder36) [inline QMigrate/io/export/to_obspy.py:105]:
```suggestion
    event_info = pd.read_csv(event_file).iloc[0]
    event_uid = str(event_info)
```

File: QMigrate/signal/local_mag/magnitude.py (hunk 2 above)
Comment (hemmelig) [reply QMigrate/signal/local_mag/magnitude.py]: Resolved.

**Pull request merged**